src/librustc_typeck/error_codes.rs

```diff
@@ -2425,6 +2425,85 @@ struct Bar<S, T> { x: Foo<S, T> }
 ```
 "##,
 
+E0307: r##"
+This error indicates that the `self` parameter in a method has an invalid
+"reciever type".
+
+Methods take a special first parameter, of which there are three variants:
+`self`, `&self`, and `&mut self`. These are syntactic sugar for
+`self: Self`, `self: &Self`, and `self: &mut Self` respectively. The type
+`Self` acts as an alias to the type of the current trait implementer, or
+"receiver type". Besides the already mentioned `Self`, `&Self` and
+`&mut Self` valid receiver types, the following are also valid:
+`self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, and `self: Pin<P>`
+(where P is one of the previous types except `Self`).
```

Comment:
Hmm; is it sufficiently clear here that `self: Box<Self>` is not meant literally (e.g. `self: Box<Foo>` is fine as well)?

Reply:
This copy is taken from the current diagnostic, so if we should change it too. What about "a type that _can be_ Self". Then again, the extended explanation should probably go in the book and linked from the error code index.

Reply:
Maybe we could instead add a sentence: "Note that `Self` is not meant literally here. Any type which resolves to the underlying implementing type is fine."?

Reply:
"Note that `Self` can also be the underlying implementing type, like `Foo` in the following example."

Reply:
👍 -- better to connect to an example.